federatedscope/core/sampler.py

```diff
@@ -0,0 +1,124 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+
+class Sampler(ABC):
+    """
+    The strategies of sampling clients for each training round
+
+    Arguments:
+        client_state: a dict to manager the state of clients (idle or busy)
+    """
+    def __init__(self, client_num):
+        self.client_state = np.asarray([1] * (client_num + 1))
+        # Set the state of server (index=0) to 'working'
+        self.client_state[0] = 0
+
+    @abstractmethod
+    def sample(self, size):
+        raise NotImplementedError
+
+    def change_state(self, indices, state):
+        """
+        To modify the state of clients (idle or working)
+        """
+        if state == 'idle':
+            self.client_state[indices] = 1
+        elif state == 'working':
+            self.client_state[indices] = 0
+        else:
+            raise ValueError(
+                f"The state of client should be 'idle' or 'working', but got {state}"
+            )
+
+
+class UniformSampler(Sampler):
+    """
+    To uniformly sample the clients from all the idle clients
+    """
+    def __init__(self, client_num):
+        super(UniformSampler, self).__init__(client_num)
+
+    def sample(self, size):
+        """
+        To sample clients
+        """
+        idle_clients = np.nonzero(self.client_state)[0]
+        sampled_clients = np.random.choice(idle_clients,
+                                           size=size,
+                                           replace=False).tolist()
```

Comment:
Is it possible to make this argument configurable?

Reply:
Do you mean `replace`? It should be set to `False` to avoid sampling repeated clients.